### DISCUSSION TITLE
Pattern matching case classes and companion object

### DISCUSSION THREAD

**Discussion by sofoklis** [scala/scala3#21582] [General Question] [answered]
hi,
in scala 3.5, this is not giving a warning or a compiler error without any flags.

```
@main def ddd =
  val s: Example = Example(1)

  s match {
    case Example => println("Example")
  }  


case class Example(x: Int)  
```

then we get a runtime error like this:

```
[info] running ddd 
[error] scala.MatchError: Example(1) (of class Example)
[error]         at ExtentionMethods$package$.ddd(ExtentionMethods.scala:46)
[error]         at ddd.main(ExtentionMethods.scala:42)
[error]         at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
[error]         at java.base/java.lang.reflect.Method.invoke(Method.java:580)
[error] stack trace is suppressed; run last Compile / run for the full output
[error] (Compile / run) scala.MatchError: Example(1) (of class Example)
[error] Total time: 2 s, completed Sep 13, 2024, 11:20:20 AM

```
Here the pattern matches on the case class companion object, is it really of type Example? Why is it even allowed in the pattern match by the compiler? You only get a warning if you put the  compiler flag  "-Ycheck-all-patmat"

**Comment by bracevac** [comment]
Hi, this seems to be a regression in Scala 3.5. Please open an issue ticket for it. Thanks!

**Comment by dwijnand** [comment, marked as answer]
Those warnings were moved to `-Wimplausible-patterns` in 3.4.  In https://github.com/scala/scala3/pull/18093

**Reply by sofoklis** [reply]
thank u for the answer, it feels dangerous to have a flag for this behavior though.

**Reply by dwijnand** [reply]
I can definitely sympathise with that sentiment, which is why I spend lots of time on the match analysis component of the compiler (also in Scala 2).  I haven't looked hard at the PR, but apparently we approximate and can be wrong on both false positive and false negative sides, so making it opt-in seems reasonable under those circumstances, but I don't know the reasons why we fail to approximate well enough.
